Unauthorized when uploading a report into Google Drive

Occasionally user might end up with Forbidden from Google Drive when trying to upload a report into Google Drive

    2018-01-28 23:31:15,623 SEVERE [smvcj] (default task-18) null: fi.metatavu.edelphi.smvcj.SmvcRuntimeException: Virhe viessä raportin datoja.
      at fi.metatavu.edelphi.binaries.queries.QueryDataExportBinaryController.exportGoogleSpreadsheet(QueryDataExportBinaryController.java:126)
      at fi.metatavu.edelphi.binaries.queries.QueryDataExportBinaryController.process(QueryDataExportBinaryController.java:85)
      at fi.metatavu.edelphi.smvcj.dispatcher.Servlet.doService(Servlet.java:183)
      at fi.metatavu.edelphi.smvcj.dispatcher.Servlet.service(Servlet.java:95)
      at fi.metatavu.edelphi.DispatcherServlet.service(DispatcherServlet.java:22)
      at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
      at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
      at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:81)
      at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
      at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:274)
      at io.undertow.servlet.handlers.ServletInitialHandler.dispatchToPath(ServletInitialHandler.java:209)
      at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImpl(RequestDispatcherImpl.java:221)
      at io.undertow.servlet.spec.RequestDispatcherImpl.forwardImplSetup(RequestDispatcherImpl.java:147)
      at io.undertow.servlet.spec.RequestDispatcherImpl.forward(RequestDispatcherImpl.java:111)
      at fi.metatavu.edelphi.PrettyUrlFilter.doFilter(PrettyUrlFilter.java:28)
      at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
      at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
      at fi.metatavu.edelphi.CommonsFilter.doFilter(CommonsFilter.java:46)
      at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
      at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
      at fi.metatavu.edelphi.EncodingFilter.doFilter(EncodingFilter.java:25)
      at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
      at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
      at fi.metatavu.edelphi.LocaleFilter.doFilter(LocaleFilter.java:45)
      at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
      at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
      at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
      at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
      at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
      at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
      at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
      at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
      at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
      at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
      at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
      at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
      at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
      at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
      at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
      at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
      at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
      at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
      at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
      at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
      at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
      at io.undertow.server.Connectors.executeRootHandler(Connectors.java:202)
      at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:805)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
      at java.lang.Thread.run(Thread.java:748)
    Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 401 Unauthorized
      at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
      at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:113)
      at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:40)
      at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:432)
      at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
      at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
      at fi.metatavu.edelphi.utils.GoogleDriveUtils.tryInsertFile(GoogleDriveUtils.java:451)
      at fi.metatavu.edelphi.utils.GoogleDriveUtils.insertFile(GoogleDriveUtils.java:168)
      at fi.metatavu.edelphi.binaries.queries.QueryDataExportBinaryController.exportGoogleSpreadsheet(QueryDataExportBinaryController.java:118)
      ... 67 more

--- a/edelphi/src/main/java/fi/metatavu/edelphi/utils/AuthUtils.java
+++ b/edelphi/src/main/java/fi/metatavu/edelphi/utils/AuthUtils.java
@@ -144,6 +144,17 @@ public class AuthUtils {
     }
   }
 
+  /**
+   * Returns OAuth previously stored access token from HTTP session
+   * 
+   * Null is returned if token is present but does not have requested scopes or 
+   * the token has been expired
+   * 
+   * @param requestContext request context
+   * @param provider token provider
+   * @param scopes required scopes
+   * @return OAuth previously stored access token from HTTP session
+   */
   public static OAuthAccessToken getOAuthAccessToken(RequestContext requestContext, String provider, String... scopes) {
     HttpSession session = requestContext.getRequest().getSession();
     
@@ -151,20 +162,20 @@ public class AuthUtils {
     if (accessTokens != null) {
       for (OAuthAccessToken accessToken : accessTokens) {
         List<String> accessTokenScopes = accessToken.getScopes() != null ? Arrays.asList(accessToken.getScopes()) : Collections.emptyList();
-        if (scopes == null || accessTokenScopes.containsAll(Arrays.asList(scopes))) {
-      		return accessToken;
+        if (scopes == null || accessTokenScopes.containsAll(Arrays.asList(scopes)) && !isOAuthTokenExpired(accessToken)) {
+          return accessToken;
       	}
       }
     }
     
     return null;
   }
-  
+
   public static boolean isGrantedOAuthScope(RequestContext requestContext, String provider, String scope) {
     return getOAuthAccessToken(requestContext, provider, scope) != null;
   }
   
-  public static boolean isOAuthTokenExpired(OAuthAccessToken accessToken) {
+  private static boolean isOAuthTokenExpired(OAuthAccessToken accessToken) {
     return (accessToken.getExpires() != null) && (accessToken.getExpires().getTime() < System.currentTimeMillis());
   }
   

